### ISSUE TITLE
Human Pose as Compositional Tokens

### ISSUE THREAD

**Issue by gwleeee** [gwleeee/PaperReview#10]
# Human Pose as Compositional Tokens

https://github.com/gengzigang/pct

https://arxiv.org/abs/2303.11638

https://sites.google.com/view/pctpose

CVPR2023

## Abstract

- 휴먼 포즈는 관절의 좌표 혹은 히트맵으로 표현
    - 데이터 처리가 쉽지만 신체 관절간 종속성 모델링이 없어 비현실적인 포즈 추정이 허용됨

![image](https://github.com/gwleeee/PaperReview/assets/83268025/b8a7152b-94ad-40b4-8062-177c89571a9f)

- Pose as Compositional Tokens (PCT)
    - joint dependency 탐색
        - Fig 1.을 보면, 특정 토큰을 변화시켜 디코딩 하였을 때, 특정 부분의 관절만이 변화함을 알수 있음
    - 휴먼포즈를 M토큰화 (M discrete tokens)
    - PCT 정보는 인코더, 디코더를 통해 학습 및 사용 됨
- 이미지로 부터 M토큰을 분류하는 classifier
    - 분류된 M토큰을 이용하여 디코더로 포즈 복원

![image](https://github.com/gwleeee/PaperReview/assets/83268025/b5b86f03-9f43-408a-89d2-9086a7099c29)
## Pose as Compositional Tokens (PCT)

![image](https://github.com/gwleeee/PaperReview/assets/83268025/c95ee955-c8b0-4d6c-84ad-0e594d4aa231)
### 1. Learning compositional tokens

![image](https://github.com/gwleeee/PaperReview/assets/83268025/9e531e04-4a73-4192-b950-fc0e10ffb703)
- Encoder
    
    ![image](https://github.com/gwleeee/PaperReview/assets/83268025/c55a5eae-4880-4740-9628-00fd76c7acd4)    
    ![image](https://github.com/gwleeee/PaperReview/assets/83268025/f9e69944-1967-418e-a8c8-9113541ef68e)
    
    - 인코더 f_e를 통해 pose 정보를 M개의 토큰화 함
        - G는 KxD 차원의 raw pose
        - K: number of joints
        - D: joint dim
        - 각 토큰 t는 H개의 sub-structure로 구성
    - 인코더의 구조를 단계별로 보면
        1. pose의 차원을 토큰의 feature와 같은 차원(H)을 가지도록 linear projection
        2. N개의 MLP-mixer를 이용하여 서로 다른 joint의 feature를 결합
        3. 결합된 feature를 기준으로 K개의 joint를 토큰 개수(M)개로 linear projection
- Codebook
    
    ![image](https://github.com/gwleeee/PaperReview/assets/83268025/0afd8e99-c6bc-4faf-b53c-50fb9baa6ac5)
    
    ![image](https://github.com/gwleeee/PaperReview/assets/83268025/35ac359f-6bc0-4fad-9e36-a9be3802f675)
    
    - 인코더를 통해 추출된 각 토큰의 H차원의 feature를 latent embedding space로 양자화
        - nearest lookup
    - 총 V개의 코드로 구성, 각 코드는 N차원(H차원)의 값을 가짐 (논문의 표기 오류인듯 함)
        - 이 연구에서는 1024개의 코드 사용
        - 양자화에러를 최소화 하기 위해서는 입력 이미지의 화소수 만큼의 codebook 필요
    - quentized token
        
        ![image](https://github.com/gwleeee/PaperReview/assets/83268025/df95721d-3bf0-4fae-b354-e44ce7584022)
        
- Decoder
    
    ![image](https://github.com/gwleeee/PaperReview/assets/83268025/e5950077-f465-48b6-8186-288296a9a71b)
    
    - 인코더 네트워크와 구조적으로 반대
    - MLP-mixer는 하나만 사용함
- Training
    
    ![image](https://github.com/gwleeee/PaperReview/assets/83268025/52de2aff-c968-4d3b-baa6-2286e80fbe7b)
    
    - sg: stopping gredient
    - 첫번째 term은 입출력의 smooth L1 loss
    - **두번째 term은 토큰의 feature가 코드와 동일하게 인코딩 되게끔 L2 loss**

### 2. Human pose estimation

![image](https://github.com/gwleeee/PaperReview/assets/83268025/9ff25541-1672-4594-b2df-76b1f0d3d353)
학습된 코드북과 디코더를 사용하여 포즈 추론

- classification head
    
    ![image](https://github.com/gwleeee/PaperReview/assets/83268025/78cc7e39-7458-4cbf-9a7c-e642ca663e06)
    
    ![image](https://github.com/gwleeee/PaperReview/assets/83268025/4c61c86f-72f7-40cf-ba8c-cafed1d26e6f)
    
    - backbone feature를 flatten, linear projection 후 MxN 모양으로 만듬
    - 4개의 MLP-Mixer
    - logit을 통과시켜 확률 → 값 (각 값은 code)
- Training
    
    ![image](https://github.com/gwleeee/PaperReview/assets/83268025/34052e08-f81d-46cc-b25d-38cc2d12c243)
    
    - L은 GT포즈와 인코더를 통해 얻은 토큰 클래스
    
    ![image](https://github.com/gwleeee/PaperReview/assets/83268025/4ce37818-73ad-4f21-8d3b-448a60b31dc2)
    

## Experiments

1. Datasets and metrics
    - Dataset
        - 2D: COCO, MPII / occlusion: CrowdPose, OCHuman, SyncOCC, OCHuman
        - 3D: H36M
    - Evaluation metric: OKS, AP, PCKh, MPJPE
2. Implementation details
    - GT bbox가 제공되는 데이터셋 사용 학습
    - backbone: SwinTransformerV2 (imageNet-1k)
        - backbone은 고정하고 head만 학습
    - MMPose 이용 augmentation (scale, rotation, flip, dropout, color jitter)
    - AdamW + warm restart
3. COCO, MPII
    
    ![image](https://github.com/gwleeee/PaperReview/assets/83268025/17b52e4c-c0dd-4da3-9009-dd93ca4e945f)
    
    ![image](https://github.com/gwleeee/PaperReview/assets/83268025/947f08eb-7b8e-4272-a772-3d9c3044e7ef)
    
4. Occlusion
    
    ![image](https://github.com/gwleeee/PaperReview/assets/83268025/c20b7d8d-918f-4ec4-9d00-ac7b64d4a80c)
    
    ![image](https://github.com/gwleeee/PaperReview/assets/83268025/65f0715b-6ad1-4b2f-8f35-1448cd5985c8)
    
5. number of codebook entries
    
    ![image](https://github.com/gwleeee/PaperReview/assets/83268025/77bc80f4-669c-48c4-b45d-ea5742634f84)
    

![image](https://github.com/gwleeee/PaperReview/assets/83268025/338d0144-12eb-4643-9a68-b1e40eb60d96)

![image](https://github.com/gwleeee/PaperReview/assets/83268025/53d42aed-cd66-4bb9-a46c-9e631d4e487f)
